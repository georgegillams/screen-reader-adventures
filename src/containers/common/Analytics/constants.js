export const KEY = 'analytics';
export const SEND_ANALYTIC = `app/${KEY}/SEND_ANALYTIC`;
