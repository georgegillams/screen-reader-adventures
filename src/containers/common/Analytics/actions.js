import { createRoutine } from 'redux-saga-routines';

import { SEND_ANALYTIC } from './constants';

export const sendAnalytic = createRoutine(SEND_ANALYTIC);
