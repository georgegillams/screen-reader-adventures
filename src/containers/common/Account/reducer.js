import produce from 'immer';

import { logout, requestVerificationEmail } from './actions';

export const initialState = {
  loggingOut: false,
  logOutError: null,
  logOutResult: null,
  requestingVerificationEmail: false,
  requestVerificationEmailError: null,
  requestVerificationEmailResult: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case logout.REQUEST:
        draft.loggingOut = true;
        draft.logOutError = null;
        break;

      case logout.SUCCESS:
        draft.loggingOut = false;
        draft.logOutResult = payload;
        break;

      case logout.FAILURE:
        draft.loggingOut = false;
        draft.logOutError = payload;
        break;

      case requestVerificationEmail.REQUEST:
        draft.requestingVerificationEmail = true;
        draft.requestVerificationEmailError = null;
        break;

      case requestVerificationEmail.SUCCESS:
        draft.requestingVerificationEmail = false;
        draft.requestVerificationEmailResult = payload;
        break;

      case requestVerificationEmail.FAILURE:
        draft.requestingVerificationEmail = false;
        draft.requestVerificationEmailError = payload;
        break;
    }
  });

export default reducer;
