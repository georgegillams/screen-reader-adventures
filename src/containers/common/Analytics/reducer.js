import produce from 'immer';

import { sendAnalytic } from './actions';

export const initialState = {
  sending: false,
  sendError: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case sendAnalytic.TRIGGER:
        draft.analytic = payload;
        break;

      case sendAnalytic.REQUEST:
        draft.sending = true;
        draft.sendError = null;
        break;

      case sendAnalytic.SUCCESS:
        draft.sending = false;
        break;

      case sendAnalytic.FAILURE:
        draft.sending = false;
        draft.sendError = payload;
        break;
    }
  });

export default reducer;
