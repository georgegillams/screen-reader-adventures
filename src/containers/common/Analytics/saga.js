import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { selectState } from './selectors';
import { sendAnalytic } from './actions';

export function* doSendAnalytic() {
  const currentState = yield select(selectState());
  const { analytic } = currentState;
  const requestURL = apiStructure.createAnalytic.fullPath;

  try {
    yield put(sendAnalytic.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(analytic),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(sendAnalytic.failure(result));
    } else {
      yield put(sendAnalytic.success(result));
    }
  } catch (err) {
    yield put(sendAnalytic.failure(err));
  } finally {
    yield put(sendAnalytic.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(sendAnalytic.TRIGGER, doSendAnalytic);
}
