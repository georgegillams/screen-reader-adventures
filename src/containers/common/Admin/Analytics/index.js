import { memo } from 'react';

import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { load } from './actions';
import { selectState } from './selectors';
import { selectState as selectAuthenticatorState } from 'containers/common/Authenticator/selectors';
import AdminAnalytics from './Container';
import injectSaga from 'client-utils/common/redux/inject-saga';
import injectReducer from 'client-utils/common/redux/inject-reducer';

import { KEY } from './constants';
import saga from './saga';
import reducer from './reducer';

const mapStateToProps = createStructuredSelector({
  adminAnalyticsState: selectState(),
  authenticatorState: selectAuthenticatorState(),
});

export function mapDispatchToProps(dispatch) {
  return {
    load: payload => dispatch(load(payload)),
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);
const withSaga = injectSaga({ key: KEY, saga });
const withReducer = injectReducer({ key: KEY, reducer });

export default compose(withSaga, withReducer, withConnect, memo)(AdminAnalytics);
