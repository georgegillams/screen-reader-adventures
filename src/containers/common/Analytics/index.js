import { memo } from 'react';

import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { sendAnalytic } from './actions';
import { selectState as selectConsentState } from '../Consent/selectors';
import { selectState } from './selectors';
import Analytics from './Container';
import injectSaga from 'client-utils/common/redux/inject-saga';
import injectReducer from 'client-utils/common/redux/inject-reducer';

import { KEY } from './constants';
import saga from './saga';
import reducer from './reducer';

const mapStateToProps = createStructuredSelector({
  consentState: selectConsentState(),
  analyticState: selectState(),
});

export function mapDispatchToProps(dispatch) {
  return {
    sendAnalytic: payload => dispatch(sendAnalytic(payload)),
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);
const withSaga = injectSaga({ key: KEY, saga });
const withReducer = injectReducer({ key: KEY, reducer });

export default compose(withSaga, withReducer, withConnect, memo)(Analytics);
