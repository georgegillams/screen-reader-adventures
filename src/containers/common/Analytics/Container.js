import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { DebugObject } from 'components/common/DebugObject';
import { CONSENT_STATE_ALLOWED } from 'containers/common/Consent/constants';
import { detect } from 'detect-browser';
import { withRouter } from 'next/router';

const Analytics = props => {
  const [analytic, setAnalytic] = useState(null);
  const [analyticSent, setAnalyticSent] = useState(false);

  const {
    router,

    sendAnalytic,

    analyticState,
    consentState,

    className,
  } = props;

  const constructAnalytic = () => {
    if (!router) {
      return null;
    }
    let utm_medium;
    let utm_source;
    let url;
    let browser;
    let browserVersion;
    let os;

    const detectResult = detect();
    if (detectResult) {
      browser = detectResult.name;
      browserVersion = detectResult.version;
      os = detectResult.os;
    }
    if (router) {
      url = router.pathname;
      if (router.query) {
        utm_medium = router.query.utm_medium;
        utm_source = router.query.utm_source;
      }
    }
    return {
      type: 'SPA_LOAD',
      url,
      utm_source,
      utm_medium,
      browser,
      browserVersion,
      os,
    };
  };

  useEffect(() => {
    // Create the analytic data as soon as the container loads...
    const newAnalytic = constructAnalytic();
    if (!analytic) {
      setAnalytic(newAnalytic);
    }
    const analyticToSend = analytic || newAnalytic;
    // ... but wait to send it until we have consent
    if (analyticToSend && !analyticSent && consentState.cookieConsent === CONSENT_STATE_ALLOWED) {
      sendAnalytic(analyticToSend);
      setAnalyticSent(true);
    }
  }, [consentState]);

  const outerClassNames = [];

  if (className) {
    outerClassNames.push(className);
  }

  return (
    <div className={outerClassNames.join(' ')}>
      <DebugObject
        debugTitle="Analytics"
        debugObject={{
          sendAnalytic,
          analyticState,
          consentState,
          className,
        }}
      />
    </div>
  );
};

Analytics.propTypes = {
  router: PropTypes.object,
  sendAnalytic: PropTypes.func.isRequired,
  consentState: PropTypes.shape({
    cookieConsent: PropTypes.string,
  }).isRequired,
  analyticState: PropTypes.shape({
    sending: PropTypes.bool,
    sendError: PropTypes.object,
  }).isRequired,
  className: PropTypes.string,
};

Analytics.defaultProps = {
  router: null,
  className: null,
};

export default withRouter(Analytics);
