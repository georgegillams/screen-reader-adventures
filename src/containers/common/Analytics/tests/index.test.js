import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { initialState as initialConsentState } from '../../Consent/reducer';
import { initialState } from '../reducer';

import configureStore from 'client-utils/common/redux/configure-store';

import AnalyticsIndex from '../index';
import Analytics from '../Container';

describe('<Analytics />', () => {
  let store;
  const spy = jest.fn();

  beforeAll(() => {
    store = configureStore({});
  });

  it('should render correctly - index', () => {
    const { container } = render(
      <Provider store={store}>
        <AnalyticsIndex />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const { container } = render(
      <Provider store={store}>
        <Analytics
          sendAnalytic={spy}
          consentState={{
            ...initialConsentState,
          }}
          analyticState={{
            ...initialState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
