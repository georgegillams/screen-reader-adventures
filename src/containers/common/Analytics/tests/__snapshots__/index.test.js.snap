// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Analytics /> should render correctly - index 1`] = `
<div>
  <div
    class=""
  />
</div>
`;

exports[`<Analytics /> should render correctly 1`] = `
<div>
  <div
    class=""
  />
</div>
`;
